src/Symfony/Component/Security/Http/Authentication/AuthenticatorManager.php

```diff
@@ -0,0 +1,241 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authentication;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\AuthenticationEvents;
+use Symfony\Component\Security\Core\Event\AuthenticationSuccessEvent;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\Authenticator\AuthenticatorInterface;
+use Symfony\Component\Security\Http\Authenticator\InteractiveAuthenticatorInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+use Symfony\Component\Security\Http\Event\InteractiveLoginEvent;
+use Symfony\Component\Security\Http\Event\LoginFailedEvent;
+use Symfony\Component\Security\Http\Event\LoginSuccessEvent;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+use Symfony\Component\Security\Http\SecurityEvents;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ * @author Amaury Leroux de Lens <amaury@lerouxdelens.com>
+ *
+ * @experimental in 5.1
+ */
+class AuthenticatorManager implements AuthenticatorManagerInterface
+{
+    private $authenticators;
+    private $tokenStorage;
+    private $eventDispatcher;
+    private $eraseCredentials;
+    private $logger;
+    private $providerKey;
+
+    /**
+     * @param AuthenticatorInterface[] $authenticators The authenticators, with their unique providerKey as key
+     */
+    public function __construct(iterable $authenticators, TokenStorageInterface $tokenStorage, EventDispatcherInterface $eventDispatcher, string $providerKey, ?LoggerInterface $logger = null, bool $eraseCredentials = true)
+    {
+        $this->authenticators = $authenticators;
+        $this->tokenStorage = $tokenStorage;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->providerKey = $providerKey;
+        $this->logger = $logger;
+        $this->eraseCredentials = $eraseCredentials;
+    }
+
+    public function authenticateUser(UserInterface $user, AuthenticatorInterface $authenticator, Request $request): ?Response
+    {
+        // create an authenticated token for the User
+        $token = $authenticator->createAuthenticatedToken($passport = new SelfValidatingPassport($user), $this->providerKey);
```

Comment:
Hmm. So *now* that we have the cool Passport system, I realize that there may be a use-case (while manually authenticating) to pass badges. One immediate example is "remember me": if I want to manually authenticate *and* activate the remember me functionality, then I'll need the `RememberMeBadge` to be passed here.

The easiest way I can think to allow that is to add an optional `array $badges = []` 4th argument.